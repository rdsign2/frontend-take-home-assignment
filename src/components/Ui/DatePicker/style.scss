@import '~/assets/styles/configs.scss';

.DatePicker {
  display: flex;
  flex-direction: column;

  button {
    height: 100%;
    letter-spacing: -0.02em;

    img {
      transition: all 0.2s ease-out;
      width: 0.9rem;
    }

    &:active {
      img {
        transform: scale(0.8);
      }
    }
  }

  .back {
    background: $color_light_hard;
    width: 6rem;

    img {
      margin-left: -0.3rem;
    }

    &:disabled {
      background: $color_light;

      img {
        opacity: 0.3;
      }
    }
  }

  .forward {
    background: $color_light_hard;
    transform: rotate(180deg);
    width: 6rem;

    img {
      margin-right: 0.3rem;
    }
  }

  .pickerDisplay {
    flex-flow: column;
    line-height: 1rem;
    position: relative;
    width: calc(100% - 12rem);

    &:focus {
      box-shadow: inset 0 0 0.5vh rgba($color_primary_soft, 0.15);
    }

    p {
      font-size: 1.9rem;
      line-height: 1.33em;
      text-transform: capitalize;

      small {
        font-size: 0.7em;
        font-weight: normal;
        line-height: 0.7em;
      }
    }
  }

  .innerHolder {
    align-items: center;
    border: $border;
    border-radius: $border_radius_md;
    display: flex;
    height: 6rem;
    justify-content: space-between;
    margin-top: 0.5vh;
    position: relative;
    width: 100%;
  }

  .picker {
    animation: fade 0.25s ease-in-out;
    animation-iteration-count: 1;
    background: $color_light_soft;
    border: $border;
    border-radius: $border_radius_md;
    left: -1px;
    position: absolute;
    top: -1px;
    width: calc(100% + 2px);
    z-index: 2;

    .row {
      height: 6vh;

      .back,
      .forward {
        background: transparent;
      }

      .pickerDisplay {
        cursor: default;

        p {
          color: $color_primary_soft;
        }
      }
    }

    ul {
      box-shadow: 0 1vh 3vh rgba(black, 0.1);
      display: flex;
      flex-wrap: wrap;
      padding: 1vh 2vw 2.5vh;
      z-index: 1;

      li {
        border-radius: 100px;
        cursor: pointer;
        padding: 0.75vh 0;
        text-align: center;
        transition: all 0.15s ease-in-out;
        user-select: none;
        width: 33.3%;

        &:hover {
          background: $color_primary;

          p {
            color: $color_light_soft;
          }
        }

        &:active {
          background: $color_primary_soft;
          color: $color_light_soft;

          p {
            color: $color_light_soft;
          }
        }
      }

      .disabled {
        cursor: default;

        p {
          color: $color_dark_soft;
        }

        &:hover {
          background: inherit;

          p {
            color: $color_dark_soft;
          }
        }
      }

      .current {
        background: $color_primary_soft;

        p {
          color: $color_light_soft;
        }
      }
    }
  }
}

// @media all and (min-width: 280px) and (max-width: 679px) and (orientation: portrait) {
//   .datePicker {
//     button {
//       img {
//         width: calc(((#{$space} / 35) + 7px) * 0.75);
//       }
//     }

//     .back,
//     .forward {
//       flex: 15;
//     }

//     .pickerDisplay {
//       flex: 70;

//       p {
//         font-size: calc(#{$p} * 1.4);
//       }
//     }

//     .innerHolder {
//       height: 7vh;
//     }

//     .picker {
//       position: absolute;

//       .row {
//         height: 7vh;
//       }

//       ul {
//         padding: 2vh 10vw 3.5vh;

//         li {
//           width: 25%;
//           padding: 1.5vh 0;
//         }
//       }
//     }
//   }
// }

// @media all and (min-height: 280px) and (max-height: 679px) and (orientation: landscape) {
//   .datePicker {
//     .back,
//     .forward {
//       flex: 15;
//     }

//     .innerHolder {
//       height: 7vw;
//     }

//     .picker {
//       .row {
//         height: 7vw;
//       }

//       ul {
//         li {
//           padding: 0.75vw 0;
//         }
//       }
//     }
//   }
// }
