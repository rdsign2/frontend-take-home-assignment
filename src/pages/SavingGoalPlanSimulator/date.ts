import moment from 'moment';

export const months = moment.months();
export const currentYear = new Date().getFullYear();
export const currentMonth = new Date().toLocaleString('en-US', {
  month: 'long'
});
export const formatter = new Intl.NumberFormat('en-US', {
  useGrouping: true,
  minimumFractionDigits: 0,
  maximumFractionDigits: 0
});
export const minMonthGoalPeriod = 1;
export const minStartingMonthIndex =
  months.indexOf(currentMonth) + minMonthGoalPeriod;
export const minStartingMonth = months[minStartingMonthIndex];
export const disabledMonths = months.filter((e, index) => {
  if (index < months.indexOf(minStartingMonth)) {
    return e;
  }
});

export const changeDate = (
  action: 'prev' | 'next',
  month: string,
  year: number
) => {
  let newMonth = month;
  let newYear = year;
  const currMonthIndex = months.indexOf(month);
  if (action === 'next') {
    month === 'December'
      ? ((newMonth = 'January'), (newYear += 1))
      : (newMonth = months[currMonthIndex + 1]);
  } else {
    month === minStartingMonth && year === currentYear
      ? null
      : month === 'January'
      ? ((newMonth = 'December'), (newYear -= 1))
      : (newMonth = months[currMonthIndex - 1]);
  }
  return {
    newMonth,
    newYear
  };
};

export const pickDate = (pickedMonth: number, year: number) => {
  const currMonthIndex = months.indexOf(currentMonth);
  if (pickedMonth <= currMonthIndex && year === currentYear) {
    return;
  }
  return months[pickedMonth];
};
