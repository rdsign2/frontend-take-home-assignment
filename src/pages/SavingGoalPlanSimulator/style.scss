@import '~/assets/styles/configs.scss';

.SavingGoalPlanSimulator {
  width: 62rem;

  h3 {
    color: $color_primary;
    font-weight: 400;
    line-height: 1em;
    padding: 4rem 0;
    text-align: center;

    strong {
      font-weight: bold;
    }
  }

  .goalPlanCalculator {
    background-color: $color_light_soft;
    border-radius: $border_radius_lg;
    border: $border;
    padding: $padding_lg;
    margin-bottom: 4rem;

    .heading {
      margin-bottom: $padding_lg;

      .IconHouse {
        height: auto;
        margin-bottom: 0.5rem;
        width: 3.6rem;
      }

      h1 {
        font-size: 3.2rem;
        margin-bottom: 0.5rem;
      }
    }

    .inputs {
      justify-content: space-between;
      width: 100%;

      & .Input {
        width: 47.5%;
      }

      & .DatePicker {
        width: 47.5%;
      }
    }

    .summary {
      align-items: center;
      border: $border;
      border-radius: $border_radius_md;
      box-shadow: 0 0.15rem 0.6rem rgba($color_dark_hard, 0.05);
      margin: $padding_md 0;

      :nth-child(1).row {
        align-items: center;
        justify-content: space-between;
        padding: $padding_lg;
      }

      .label {
        font-size: 1.8rem;
        font-weight: 500;
      }

      .amount {
        color: $color_primary_soft;
        font-size: 4rem;
        font-weight: 400;
        letter-spacing: -0.04em;

        strong {
          font-weight: 500;
        }
      }

      :nth-child(2).row {
        padding: 2rem $padding_lg;
        background: $color_light;
        border-radius: 0 0 $border_radius_md $border_radius_md;
      }

      .details {
        p {
          font-size: 1.4rem;

          strong {
            font-weight: 600;
          }
        }
      }
    }

    .Button {
      margin-bottom: 1rem;
      width: 80%;
    }
  }
}

@media all and (min-width: 260px) and (max-width: 600px) and (orientation: portrait) {
  .SavingGoalPlanSimulator {
    align-self: flex-end;
    height: 100%;
    width: 100%;

    .goalPlanCalculator {
      border: 0;
      border-radius: $border_radius_lg_mob $border_radius_lg_mob 0 0;
      box-shadow: 0 0.6rem 2.4rem rgba($color_dark_hard, 0.2);
      height: -webkit-fill-available;
      // height: calc(100% - 10rem);
      padding: $padding_lg_mob;
      margin-bottom: 0;

      .heading {
        margin-bottom: $padding_md_mob;
      }

      .inputs {
        & .Input {
          width: 100%;
          margin-bottom: $padding_md_mob;
        }

        & .DatePicker {
          width: 100%;
        }
      }

      .summary {
        border-radius: $border_radius_md;
        margin: $padding_md_mob 0;

        :nth-child(1).row {
          padding: $padding_md_mob;
        }

        .label {
          span {
            display: none;
          }
        }

        .amount {
          font-size: 3.2rem;
        }

        :nth-child(2).row {
          padding: $padding_md_mob;
          border-radius: 0 0 $border_radius_md $border_radius_md;
        }

        .details {
          p {
            font-size: 1.3rem;
          }
        }
      }

      .Button {
        margin-bottom: $padding_lg_mob;
        width: 100%;
      }
    }
  }
}
