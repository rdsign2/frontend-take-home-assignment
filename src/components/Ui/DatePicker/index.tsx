import * as React from 'react';
import './style.scss';

interface Props {
  currentMonth: string;
  currentYear: number;
  disableds: string[];
  nextMonth: () => void;
  nextYear: () => void;
  minStart: string;
  months: string[];
  pickedDate: (monthIndex: number) => number | void;
  previousMonth: () => void;
  previousYear: () => void;
  stateMonth: string;
  stateYear: number;
  validateYear: () => void;
}

interface Event {
  key: string;
}

const DatePicker: React.FC<Props> = props => {
  const [picker, setPicker] = React.useState<boolean>(false);

  const handlePickDate = (pickState: boolean) => {
    setPicker(pickState);
    props.validateYear();
  };

  const handleButton = (type: string) => {
    type === 'previous' ? props.previousMonth() : props.nextMonth();
  };

  const handleKeyPress = (event: Event) => {
    if (event.key === 'ArrowLeft') {
      handleButton('previous');
    }
    if (event.key === 'ArrowRight') {
      handleButton('next');
    }
  };

  const list = props.months.map((m, index) => (
    <li
      key={index}
      onClick={() => (props.pickedDate(index), handlePickDate(false))}
      className={
        props.disableds.includes(m) && props.stateYear === props.currentYear
          ? 'disabled'
          : m === props.stateMonth
          ? 'current'
          : null
      }
    >
      <p>{m.substring(3, 0)}</p>
    </li>
  ));

  return (
    <div className="DatePicker">
      <label>Reach goal by</label>
      {/* DISPLAY */}
      <span className="innerHolder">
        <button
          className="back"
          disabled={
            props.currentYear === props.stateYear &&
            props.minStart === props.stateMonth
              ? true
              : false
          }
          onClick={() => handleButton('previous')}
          onKeyDown={handleKeyPress}
        >
          <img src={require('~/assets/icons/arrow.svg')} />
        </button>
        <button
          className="pickerDisplay"
          onClick={() => handlePickDate(!picker)}
          onKeyDown={handleKeyPress}
        >
          <p>
            {props.stateMonth}
            <small>
              <br />
              {props.stateYear}
            </small>
          </p>
        </button>
        <button
          className="forward"
          onClick={() => handleButton('next')}
          onKeyDown={handleKeyPress}
        >
          <img src={require('~/assets/icons/arrow.svg')} />
        </button>
        {/* PICKER */}
        {picker ? (
          <div className="picker" onMouseLeave={() => handlePickDate(false)}>
            <div className="row">
              <button className="back" onClick={props.previousYear}>
                <img src={require('~/assets/icons/arrow.svg')} />
              </button>
              <button
                className="pickerDisplay"
                onClick={() => handlePickDate(!picker)}
              >
                <p> {props.stateYear} </p>
              </button>
              <button className="forward" onClick={props.nextYear}>
                <img src={require('~/assets/icons/arrow.svg')} />
              </button>
            </div>
            <ul className="pickerMonths">{list}</ul>
          </div>
        ) : null}
      </span>
    </div>
  );
};

export default DatePicker;
