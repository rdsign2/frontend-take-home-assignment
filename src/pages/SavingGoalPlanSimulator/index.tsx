import * as React from 'react';
import moment from 'moment';
import Button from '../../components/Ui/Button';
import DatePicker from '../../components/Ui/DatePicker';
import Input from '../../components/Ui/Input';
import IconHouse from '../../components/SVGs/Icons/IconHouse';
import './style.scss';

const months = moment.months();
const currentYear = new Date().getFullYear();
const currentMonth = new Date().toLocaleString('en-US', { month: 'long' });
const formatter = new Intl.NumberFormat('en-US', {
  useGrouping: true,
  minimumFractionDigits: 0,
  maximumFractionDigits: 0
});
const minMonthGoalPeriod = 1;
const minStartingMonthIndex = months.indexOf(currentMonth) + minMonthGoalPeriod;
const minStartingMonth = months[minStartingMonthIndex];

const SavingGoalPlanSimulator: React.FC = () => {
  const [month, setMonth] = React.useState<string>(minStartingMonth);
  const [year, setYear] = React.useState<number>(currentYear);
  const [disabledMonths, setDisabledMonths] = React.useState<string[]>([]);
  const [deposits, setDeposits] = React.useState<number>(minMonthGoalPeriod);
  const [value, setValue] = React.useState<number>(0);
  const [installment, setInstallment] = React.useState<number>(0);

  React.useEffect(() => {
    setDisabledMonths(
      months.filter((e, index) => {
        if (index < months.indexOf(month)) {
          return e;
        }
      })
    );
  }, []);

  React.useEffect(() => {
    setDeposits(
      moment([year, months.indexOf(month), 1]).diff(
        moment([currentYear, months.indexOf(currentMonth), 1]),
        'months',
        true
      )
    );
    setInstallment(value / deposits);
    return () => {
      setDeposits(minMonthGoalPeriod);
      setInstallment(0);
    };
  }, [deposits, month, value, year]);

  const handleNextMonth = () => {
    const currMonthIndex = months.indexOf(month);
    month === 'December'
      ? (setMonth('January'), setYear(year + 1))
      : setMonth(months[currMonthIndex + 1]);
  };

  const handlePreviousMonth = () => {
    const currMonthIndex = months.indexOf(month);
    month === minStartingMonth && year === currentYear
      ? null
      : month === 'January'
      ? (setMonth('December'), setYear(year - 1))
      : setMonth(months[currMonthIndex - 1]);
  };

  const handlePickedDate = (pickedMonth: number) => {
    const currMonthIndex = months.indexOf(currentMonth);
    pickedMonth <= currMonthIndex && year === currentYear
      ? null
      : setMonth(months[pickedMonth]);
  };

  const summaryDetails = {
    empty: 'Insert total amount and choose a date.',
    deposits: ` ${deposits} ${deposits <= 1 ? 'deposit' : 'monthly deposits'} `,
    value: ` $${formatter.format(value)} `,
    when: ` ${month} ${year}.`
  };

  return (
    <section className="SavingGoalPlanSimulator">
      <h3 className="tp">
        Let&apos;s plan your <strong>saving goal.</strong>
      </h3>
      <div className="goalPlanCalculator">
        <div className="heading">
          <IconHouse color="#657786" />
          <h1>Buy a house</h1>
          <p>Saving goal</p>
        </div>
        <div className="inputs row">
          <Input valueSetter={e => setValue(e)} type="number" />
          <DatePicker
            currentMonth={currentMonth}
            currentYear={currentYear}
            disableds={disabledMonths}
            minStart={minStartingMonth}
            months={months}
            nextMonth={() => handleNextMonth()}
            nextYear={() => setYear(year + 1)}
            pickedDate={(pickedMonth: number) => handlePickedDate(pickedMonth)}
            previousMonth={() => handlePreviousMonth()}
            previousYear={() =>
              setYear(year === currentYear ? currentYear : year - 1)
            }
            stateMonth={month}
            stateYear={year}
            validateYear={() =>
              months.indexOf(month) < minStartingMonthIndex &&
              year === currentYear
                ? setYear(year + 1)
                : null
            }
          />
        </div>
        <div className="summary">
          <div className="row">
            <p className="label">
              Monthly
              <span> amount</span>
            </p>
            <h2 className="amount tp">
              $<strong>{formatter.format(installment)}</strong>
            </h2>
          </div>
          <div className="details row">
            {value === 0 ? (
              <p>{summaryDetails.empty}</p>
            ) : (
              <p>
                You&apos;re planning
                <strong>{summaryDetails.deposits}</strong>
                to reach your
                <strong>{summaryDetails.value}</strong>
                goal by
                <strong>{summaryDetails.when}</strong>
              </p>
            )}
          </div>
        </div>
        <Button enable={value} />
      </div>
    </section>
  );
};

export default SavingGoalPlanSimulator;
