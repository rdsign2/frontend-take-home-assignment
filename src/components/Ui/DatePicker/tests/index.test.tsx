import * as React from 'react';
import { shallow } from 'enzyme';
import DatePicker from '../index';

import { months } from '../../../../pages/SavingGoalPlanSimulator/date';

const mockData = {
  currentMonth: 'August',
  currentYear: 2026
};

const component = shallow(
  <DatePicker
    currentMonth={mockData.currentMonth}
    currentYear={mockData.currentYear}
    disableds={[]}
    minStart={'January'}
    months={months}
    nextMonth={jest.fn()}
    nextYear={jest.fn()}
    pickedDate={jest.fn()}
    previousMonth={jest.fn()}
    previousYear={jest.fn()}
    stateMonth={mockData.currentMonth}
    stateYear={mockData.currentYear}
    validateYear={jest.fn()}
  />
);

describe('DatePicker', () => {
  it('renders', () => {
    expect(component.exists()).toEqual(true);
  });

  it(`Picker display date should be 'August2026'`, () => {
    expect(component.find('.pickerDisplay').text()).toContain(
      `${mockData.currentMonth + mockData.currentYear}`
    );
  });

  it('snapshotShot', () => {
    expect(component.getElements()).toMatchSnapshot();
  });
});
