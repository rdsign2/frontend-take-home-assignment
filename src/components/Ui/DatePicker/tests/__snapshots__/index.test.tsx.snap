// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker snapshotShot 1`] = `
Array [
  <div
    className="DatePicker"
  >
    <label>
      Reach goal by
    </label>
    <span
      className="innerHolder"
    >
      <button
        className="back"
        disabled={false}
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <IconArrow
          color="#657786"
        />
      </button>
      <button
        className="pickerDisplay"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <p>
          August
          <small>
            <br />
            2026
          </small>
        </p>
      </button>
      <button
        className="forward"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <IconArrow
          color="#657786"
        />
      </button>
    </span>
  </div>,
]
`;
